### PR TITLE
Automatically generate outlines

### PR DESCRIPTION
**Pull request by leifp**
This PR is just for comments.  Assuming you like the idea / interface / impl, I can add tests.

### CODE REVIEW

**Comment by evidens** [comment]
This is a great addition. One or two comments about style but with a few tests I'd be happy to merge this in

File: gen_outline.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+                       help="Process each line of JSON file separately")
+    group.add_argument('-c', '--collection', type=str, default=None,
+                       help="Key in JSON of array to process", metavar="KEY")
+
+    return parser
+
+def main():
+    parser = init_parser()
+    args = parser.parse_args()
+    outline = make_outline(args)
```
Comment (evidens) [inline gen_outline.py]: Unless you think there would be a lot of variance in the arguments, I would rather see these as separate arguments (you could use `*args` to remove redundant calls). It makes it clearer what's being used in make_outline.

File: gen_outline.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    for d in iterator:
+        for path in key_paths(d):
+            key_map[tuple(path)] = True
+    return key_map
+
+def path_join(path, sep='.'):
+    return sep.join(str(k) for k in path)
+
+def key_map_to_list(key_map):
+    # We convert to strings *after* sorting to that array indices come out
```
Comment (evidens) [inline gen_outline.py]: spelling: *s*o that…

**Comment by leifp** [comment]
OK, I'll make that change and write a couple tests.  You may see the author change, since I was mistakenly using my work email.

**Comment by leifp** [comment]
OK, PR updated.  One change I made to support my use case is to handle arrays of different lengths.  You should decide whether this behavior will break existing usage.  The previous behavior was

1. Throw an IndexError if the data array was shorter than the outline spec
2. Silently ignore elements if the data array was longer than the outline spec

I changed the behavior of case 1 to return None; case 2 remains the same.

Also, I have not tested with Python 3.

**Comment by evidens** [comment]
I think you're correct with the IndexError not making sense. I never really used that functionality so your additions make sense.

Thanks for the contributions and tests. I haven't tested this on Python 3 either. For now it's fine, if someone requests it, I'll look into the issues that might arise.

**Pull request merged**